### ISSUE TITLE
Use generic to type the name of a slice

### ISSUE THREAD

**Issue by tgouala** [reduxjs/redux-toolkit#341]
Specifying a generic type for the property `name` of a slice would help typescript to deduce the type of objects built from slice.

As a practical example, I would like to put all my action creators inside an object I can then type and pass to Thunk extra arguments, or a react-redux mapDispatchToProps. This looks like this :
```ts
const clientSlice = createSlice({name:'clients', reducers: {get: (state) => {}}});
const categorySlice = createSlice({name:'category', reducers: {save: (state) => {}}});
const uiSlice = createSlice({name:'ui', reducers: {goToHome: (state) => {}}});

export const actionCreators = {
   [clientSlice.name]: {...clientSlice.actions},
   [categorySlice.name]: {...categorySlice.actions},
   [uiSlice.name]: {...uiSlice.actions}
}

export type ActionCreators = typeof actionCreators
```
If name was a generic extending string, I could have autocompletion in code like 
```ts
const connectMyComponent = (dispatch: Dispatch) : MapDispatchToProp => (actionCreators : ActionCreators) => ({
   getClients: dispatch(actionCreators.clients.get()),
   saveCategory: dispatch(actionCreators.category.save()),
   changePage: dispatch(actionCreators.ui.goToHome())
})
```

Unfortunately, this won't work since `clientSlice.name` is typed `string`.

There is a workaround you can see below, but integrating a generic type would be quite easy and straightforward (I can do a PR if wanted). 

**Workaround :**
It's possible to wrap the function createSlice like describe [here in rtk doc](https://redux-toolkit.js.org/usage/usage-with-typescript#wrapping-createslice) and return an object with a type for the property `name` :

```js
const createGenericSlice = <
  Name extends string
  T,
  Reducers extends SliceCaseReducers<T>
>({
  name = "",
  initialState,
  reducers
}: {
  name: Name
  initialState: T
  reducers: ValidateSliceCaseReducers<T, Reducers>
}) => {
  const slice = createSlice({
    name,
    initialState,
    reducers: {
      ...reducers
    }
  });

  return slice as typeof slice & {name: Name};
}

const clientSlice = createGenericSlice({
  name: 'clients',
  initialState: { status: 'loading' } as GenericState<string>,
  reducers: {
    {get: (state) => {}}
  }
})
```

then `typeof clientSlice ['name']` will be the literal 'clients'.

**Comment by tgouala** [comment]
If I add the parameter `extraReducers` to `createSlice` options it gets even more tricky since I need to use the rtk type `NoInfer<T>` which is not exported. Right now I redefine this type as shown below :
```ts
import {
  ValidateSliceCaseReducers,
  SliceCaseReducers,
  createSlice as rtkCreateSlice,
  CaseReducers,
  ActionReducerMapBuilder,
} from '@reduxjs/toolkit';

type NoInfer<T> = [T][T extends any ? 0 : never];

export const createSlice = <State, N extends string, Reducers extends SliceCaseReducers<State>>({
  name,
  initialState,
  reducers,
  extraReducers,
}: {
  name: N;
  initialState: State;
  reducers: ValidateSliceCaseReducers<State, Reducers>;
  extraReducers?:
    | CaseReducers<NoInfer<State>, any>
    | ((builder: ActionReducerMapBuilder<NoInfer<State>>) => void);
}) => {
  const slice = rtkCreateSlice({
    name,
    initialState,
    reducers: {
      ...reducers,
    },
    extraReducers: { ...extraReducers },
  });

  return slice as typeof slice & { name: N };
};
```

**Comment by phryneas** [comment]
I have [a branch where I experimented with phantom-types for action creators & actions](https://github.com/phryneas/redux-toolkit/commit/0c91d40f5f6387621014c0a54074b987c942dea7) where I actually added the `name` option as a generic parameter.

From that experiment, I totally know that this is possible, but it comes at the cost of adding a third generic parameter to the already-too-complex type definition of `createSlice`, and this is something I try to be conservative with. 
So I see the actual use case and appeal and want to say *let's do this*, but on the other hand I'm a bit hesitant to do this without a *massive* benefit.
@markerikson what do you think? The use case is valid and it's maybe 5 additional lines of code, *but* it's a third generic argument to `createSlice`.

**Comment by markerikson** [comment]
My main questions are:

- What are the actual net benefits?
- Does this have any bearing on how we generate action types, or is that still "concat gives `string`, tough"?
- How much more complexity does it add?

**Comment by phryneas** [comment]
The change suggested here: 
* `slice.name` is typed to exactly what `options.name` was, instead of `string`. 
* this would not touch the typing of `actionCreator.type` or `action.type` at all.
* complexity would be relatively low (about 5 more - not very complex - LOC), **but** it would add a third argument to the `createSlice` generic.

<hr>

The other thing I linked above would in addition to that introduce actions as phantom types, which was just an experiment I myself am not 100% sure if I wanted to do it, but as I already mentioned it, I'll explain it nonethless :laughing: 
Here's what it does:

While it does not correctly type `action.type` (because that would still be impossible), it would add two additional "phantom" (just existing in the type, not in the real object) properties (indexed by symbols so nobody gets the idea of actually using them), containing the slice name and the reducer's name as separate properties.
So for a slice `"test"` with a reducer named `"increment"` you get a `SlicePayloadAction<"test", "increment", { payload: number; type: string;}>` that in JS actually just is `{ payload: number; type: string;}`, but can be used as a restriction for `dispatch` and TS would honor it.
This has no additional value for the "discriminating union" switch/case statement though, as it cannot be tested in runtime - the only value I could see is to restict arguments to `dispatch`, so personally I'd never use it 🤷

**Comment by tgouala** [comment]
> My main questions are:
> 
> * What are the actual net benefits?
> * Does this have any bearing on how we generate action types, or is that still "concat gives `string`, tough"?
> * How much more complexity does it add?

* Benefits: Slice name property is properly typed and can be reused in definition of other types (action creators, reducers etc...). It also highlights the fact that name is a property which should not change.
* Complexity: it's roughly 5 words addition to the code base, I have a PR ready on my computer if you are interested. It indeed adds a generic type to the slice type, but slice type is already very complex, and is usually inferred by typescript when we call `createSlice`

There is workarounds for my case, but let me know if you are ever willing to integrate this directly in RTK.

**Comment by markerikson** [comment]
Sure, put up a PR and let's look at it.  Would help if you could include some tests that show the before and after results.

**Comment by markerikson** [comment]
The PR was merged in.  It'll get included in 1.3.0.  I've got several other things I want in there first, so keep an eye on when that's ready.
